_posts/2017-02-08-Rust-1.15.1.md

```diff
@@ -0,0 +1,124 @@
+    let slice1: &mut [_] = v_iter.as_mut_slice();
+    let slice2: &mut [_] = v_iter.as_mut_slice();
+    slice1[0] = 1;
+    slice2[0] = 2;
+}
+```
+
+Here both `slice1` and `slice2` are referencing the same mutable slice. Also
+notice that the iterator they are created from, `v_iter` is not declared
+mutable, which is a good indication something fishy is going on.
+
+The [solution] here is trivial, just change `&self` to `&mut self`:
+
+```rust
+pub fn as_mut_slice(&mut self) -> &mut [T] {
+    unsafe {
+        slice::from_raw_parts_mut(self.ptr as *mut T, self.len())
+    }
+}
+```
+
+[solution]: https://github.com/rust-lang/rust/pull/39466
+
+With that, proper uniqueness invariants are maintained, only one mutable slice
+can be created at a time, and `v_iter` must be declared mutable in order to pull
+out a mutable slice.
+
+So we made that change, and we're releasing a fix. In Rust we take pride in not
+breaking APIs, but since this is a new, minor feature, and the present
+implementation is spectacularly unsound, we decided to go ahead and release the
+fix immediately, hopefully before too many codebases pick it up - that is, we
+don't consider this a breaking change that requires a careful transition, but a
+necessary bug fix. For more about Rust's approach to ensuring stability see the
+["Stability as a Deliverable"][stab] blog post, [RFC 1122], on language
+evolution, and [RFC 1105], on library evolution (curiously, RFC 1105 does not
+actually contain any language allowing for library breakage due to soundness,
+but the intent has always been clear that Rust reserves the right to break code to
+fix soundness holes).
+
+[stab]: https://blog.rust-lang.org/2014/10/30/Stability.html
+[RFC 1122]: https://github.com/rust-lang/rfcs/blob/master/text/1122-language-semver.md
+[RFC 1105]: https://github.com/rust-lang/rfcs/blob/master/text/1105-api-evolution.md
+
+What can we learn from this? Well, it's at least good to be reminded why we
+endeavoured to write a safe language in the first place, that writing
+memory-unsafe code is not easy. There have been some ideas suggested for process
+improvements that would catch this class of bug. Probably the simplest is just
+to ensure patches that touch unsafe blocks get more scrutiny during review: this
+patch was written by an expert Rust programmer, reviewed by an expert Rust
+programmer, but it appeared trivial and admittedly was not scrutinized closely.
+In this we could learn from our sister project, [Servo], which uses a bot to
+alert reviewers to pull requests that touch unsafe blocks. There are other ideas
+for validating unsafe blocks in Rust that are more thorough, from human-language
+attestation that the block is correct to bolt-on theorem provers. At the moment
+the Rust library team doesn't have plans to make process changes based on this
+mistake, but unsafe validation in Rust will be an area of active research
+forever, and we'll continue to evaluate our options.
```

Comment:
So, this was where you said it felt awkward. I agree. I think maybe cutting out the ideas is better; keep it succinct. 

> What can we learn from this? Well, it's at least good to be reminded why we endeavored to write a safe language in the first place: writing safe code is not easy. We have some ideas for how to catch this kind of problem earlier in the development process, but we haven't made any decisions yet.

WDYT?